File: articles/hope-theory-positive-psychology/cpp/hope_indices.cpp
#include <iostream>

struct HopeIndicators {
    double agency_score;
    double pathways_score;
    double goal_clarity;
    double goal_progress;
    double wellbeing_score;
    double meaning_score;
    double stress_load;
    double obstacle_intensity;
    double social_support;
    double resource_access;
    double goal_revision_quality;
    double context_support;
};

double hope_index(const HopeIndicators& x) {
    return (x.agency_score + x.pathways_score) / 2.0;
}

double context_support_index(const HopeIndicators& x) {
    return (x.social_support + x.resource_access + x.context_support) / 3.0;
}

double net_pathway_context(const HopeIndicators& x) {
    return x.pathways_score + context_support_index(x) + x.goal_revision_quality - x.obstacle_intensity;
}

double net_future_orientation(const HopeIndicators& x) {
    return x.agency_score
         + x.pathways_score
         + x.goal_clarity
         + x.goal_progress
         + x.meaning_score
         + context_support_index(x)
         - x.stress_load
         - x.obstacle_intensity;
}

int main() {
    HopeIndicators example{7.1, 7.0, 7.3, 6.8, 7.1, 7.4, 3.5, 3.7, 7.2, 6.8, 6.6, 7.0};

    std::cout << "Hope index: " << hope_index(example) << "\n";
    std::cout << "Context support index: " << context_support_index(example) << "\n";
    std::cout << "Net pathway context: " << net_pathway_context(example) << "\n";
    std::cout << "Net future orientation: " << net_future_orientation(example) << "\n";

    return 0;
}

File: articles/hope-theory-positive-psychology/rust/hope_indices.rs
struct HopeIndicators {
    agency_score: f64,
    pathways_score: f64,
    goal_clarity: f64,
    goal_progress: f64,
    wellbeing_score: f64,
    meaning_score: f64,
    stress_load: f64,
    obstacle_intensity: f64,
    social_support: f64,
    resource_access: f64,
    goal_revision_quality: f64,
    context_support: f64,
}

fn hope_index(x: &HopeIndicators) -> f64 {
    (x.agency_score + x.pathways_score) / 2.0
}

fn context_support_index(x: &HopeIndicators) -> f64 {
    (x.social_support + x.resource_access + x.context_support) / 3.0
}

fn net_pathway_context(x: &HopeIndicators) -> f64 {
    x.pathways_score + context_support_index(x) + x.goal_revision_quality - x.obstacle_intensity
}

fn net_future_orientation(x: &HopeIndicators) -> f64 {
    x.agency_score
        + x.pathways_score
        + x.goal_clarity
        + x.goal_progress
        + x.meaning_score
        + context_support_index(x)
        - x.stress_load
        - x.obstacle_intensity
}

fn main() {
    let example = HopeIndicators {
        agency_score: 7.1,
        pathways_score: 7.0,
        goal_clarity: 7.3,
        goal_progress: 6.8,
        wellbeing_score: 7.1,
        meaning_score: 7.4,
        stress_load: 3.5,
        obstacle_intensity: 3.7,
        social_support: 7.2,
        resource_access: 6.8,
        goal_revision_quality: 6.6,
        context_support: 7.0,
    };

    println!("Hope index: {:.3}", hope_index(&example));
    println!("Context support index: {:.3}", context_support_index(&example));
    println!("Net pathway context: {:.3}", net_pathway_context(&example));
    println!("Net future orientation: {:.3}", net_future_orientation(&example));
}
